File: elixir/ltx/lib/interplanet_ltx/registers.ex
# registers.ex — LTX v1.1 question/action registers (Epic 72).
# Mirrors typescript/ltx/src/registers.ts (Story 69.1 —
# LTX-SPECIFICATION.md §9/§10, LTX-SECURITY.md §9.5/§9.6).
#
# Registers are not mutable state: they are deterministic reductions over
# the signed, append-only audit log. Entries are ordered by
# (timestamp, nodeId, seq) and object-level conflicts resolve by highest
# object version, then lexicographically lowest editor nodeId (§8.2).

defmodule InterplanetLtx.Registers do
  @moduledoc """
  Signed register entries (string-keyed maps with envelope
  `entryId/sessionId/nodeId/seq/type/content/timestamp/sig`), deterministic
  ordering, question/action reducers, and the Merkle entries root.
  """

  alias InterplanetLtx.Security

  @entry_prefix %{
    "question" => "QST",
    "question_response" => "QST",
    "action" => "ACT",
    "action_update" => "ACT",
    "amendment" => "AMD",
    "state_transition" => "STA",
    "merge_snapshot" => "MRG",
    "decision" => "DEC"
  }

  @action_statuses ["PROPOSED", "ACCEPTED", "REJECTED", "DONE"]

  # ── Entry creation and verification ─────────────────────────────────────────

  @doc """
  Create a signed register entry (LTX-SECURITY.md §9.5). The Ed25519
  signature covers the canonical JSON of the entry without `sig`.
  Required opts: `session_id`, `node_id`, `seq`, `timestamp`,
  `private_key_b64`; optional `entry_id`.
  """
  def create_register_entry(type, content, opts) do
    entry_id =
      Keyword.get(opts, :entry_id) ||
        "#{@entry_prefix[type]}-#{Keyword.fetch!(opts, :node_id)}-#{Keyword.fetch!(opts, :seq)}"

    unsigned = %{
      "entryId" => entry_id,
      "sessionId" => Keyword.fetch!(opts, :session_id),
      "nodeId" => Keyword.fetch!(opts, :node_id),
      "seq" => Keyword.fetch!(opts, :seq),
      "type" => type,
      "content" => content,
      "timestamp" => Keyword.fetch!(opts, :timestamp)
    }

    priv_raw = unb64u(Keyword.fetch!(opts, :private_key_b64))
    sig = :crypto.sign(:eddsa, :none, Security.canonical_json(unsigned), [priv_raw, :ed25519])
    Map.put(unsigned, "sig", b64u(sig))
  end

  @doc """
  Verify a register entry signature against a key cache mapping the entry's
  nodeId to its NIK. Returns `%{valid: bool, reason: string?}`.
  """
  def verify_register_entry(entry, key_cache) do
    sig = is_map(entry) && entry["sig"]

    cond do
      !sig ->
        %{valid: false, reason: "missing_sig"}

      true ->
        case key_cache[entry["nodeId"]] do
          nil ->
            %{valid: false, reason: "key_not_in_cache"}

          nik ->
            unsigned = Map.delete(entry, "sig")
            pub_raw = unb64u(nik["publicKey"])

            ok =
              :crypto.verify(
                :eddsa, :none,
                Security.canonical_json(unsigned),
                unb64u(sig),
                [pub_raw, :ed25519]
              )

            if ok, do: %{valid: true}, else: %{valid: false, reason: "signature_invalid"}
        end
    end
  end

  # ── Deterministic ordering (LTX-SPECIFICATION.md §8.2) ──────────────────────

  @doc "Total order key: (timestamp, nodeId, seq)."
  def entry_order_key(e), do: {e["timestamp"], e["nodeId"], e["seq"]}

  @doc "De-duplicate by (nodeId, seq) — first occurrence wins — and sort into the §8.2 total order."
  def order_entries(entries) do
    entries
    |> Enum.reduce({[], MapSet.new()}, fn e, {acc, seen} ->
      key = {e["nodeId"], e["seq"]}

      if MapSet.member?(seen, key) do
        {acc, seen}
      else
        {[e | acc], MapSet.put(seen, key)}
      end
    end)
    |> elem(0)
    |> Enum.reverse()
    |> Enum.sort_by(&entry_order_key/1)
  end

  # ── Reducers ────────────────────────────────────────────────────────────────

  # §8.2 conflict rule: higher version wins; tie → lowest editor nodeId.
  defp wins?(incoming, current) do
    if incoming.version != current.version,
      do: incoming.version > current.version,
      else: incoming.editor < current.editor
  end

  @doc """
  Reduce question register state from log entries (LTX-SPECIFICATION.md §9.4).
  Pure: identical entry sets (in any input order) produce identical state.
  Returns `%{by_id: %{qid => state}, superseded: [entryId]}` with string-keyed
  question-state maps.
  """
  def reduce_questions(entries) do
    {by_id, _winners, superseded} =
      Enum.reduce(order_entries(entries), {%{}, %{}, []}, fn e, {by_id, winners, sup} ->
        content = e["content"] || %{}

        case e["type"] do
          "question" ->
            qid = e["entryId"]

            if Map.has_key?(by_id, qid) do
              {by_id, winners, sup ++ [e["entryId"]]}
            else
              state =
                %{
                  "qid" => qid,
                  "text" => to_string(content["text"] || ""),
                  "submitter" => e["nodeId"],
                  "status" => "OPEN",
                  "version" => 1
                }
                |> put_if(content, "urgency")
                |> put_if(content, "intendedWindow")

              winner = %{version: 1, editor: e["nodeId"], entry_id: e["entryId"]}
              {Map.put(by_id, qid, state), Map.put(winners, qid, winner), sup}
            end

          "question_response" ->
            qid = to_string(content["qid"] || "")

            case by_id[qid] do
              nil ->
                {by_id, winners, sup ++ [e["entryId"]]}

              q ->
                version = content["version"] || q["version"] + 1
                incoming = %{version: version, editor: e["nodeId"], entry_id: e["entryId"]}
                current = winners[qid]

                if current && not wins?(incoming, current) do
                  {by_id, winners, sup ++ [e["entryId"]]}
                else
                  sup =
                    if current && current.entry_id != q["qid"],
                      do: sup ++ [current.entry_id],
                      else: sup

                  status =
                    if content["status"] == "WITHDRAWN", do: "WITHDRAWN", else: "ANSWERED"

                  state =
                    q
                    |> Map.merge(%{
                      "status" => status,
                      "responder" => e["nodeId"],
                      "version" => version
                    })
                    |> put_if(content, "response")

                  {Map.put(by_id, qid, state), Map.put(winners, qid, incoming), sup}
                end
            end

          _ ->
            {by_id, winners, sup}
        end
      end)

    %{by_id: by_id, superseded: superseded}
  end

  @doc """
  Reduce action register state from log entries (LTX-SPECIFICATION.md §10.2).
  Returns `%{by_id: %{aid => state}, superseded: [entryId]}`.
  """
  def reduce_actions(entries) do
    {by_id, _winners, superseded} =
      Enum.reduce(order_entries(entries), {%{}, %{}, []}, fn e, {by_id, winners, sup} ->
        content = e["content"] || %{}

        case e["type"] do
          "action" ->
            aid = e["entryId"]

            if Map.has_key?(by_id, aid) do
              {by_id, winners, sup ++ [e["entryId"]]}
            else
              state =
                %{
                  "aid" => aid,
                  "description" => to_string(content["description"] || ""),
                  "status" => "PROPOSED",
                  "version" => 1
                }
                |> put_if(content, "owner")
                |> put_if(content, "dueTimeUTC")
                |> put_if(content, "originWindow")

              winner = %{version: 1, editor: e["nodeId"], entry_id: e["entryId"]}
              {Map.put(by_id, aid, state), Map.put(winners, aid, winner), sup}
            end

          "action_update" ->
            aid = to_string(content["aid"] || "")

            case by_id[aid] do
              nil ->
                {by_id, winners, sup ++ [e["entryId"]]}

              a ->
                version = content["version"] || a["version"] + 1
                incoming = %{version: version, editor: e["nodeId"], entry_id: e["entryId"]}
                current = winners[aid]

                if current && not wins?(incoming, current) do
                  {by_id, winners, sup ++ [e["entryId"]]}
                else
                  sup =
                    if current && current.entry_id != a["aid"],
                      do: sup ++ [current.entry_id],
                      else: sup

                  status =
                    if content["status"] in @action_statuses,
                      do: content["status"],
                      else: a["status"]

                  state =
                    a
                    |> Map.merge(%{"status" => status, "version" => version})
                    |> put_if(content, "description")
                    |> put_if(content, "owner")
                    |> put_if(content, "dueTimeUTC")

                  {Map.put(by_id, aid, state), Map.put(winners, aid, incoming), sup}
                end
            end

          _ ->
            {by_id, winners, sup}
        end
      end)

    %{by_id: by_id, superseded: superseded}
  end

  defp put_if(state, content, key) do
    if Map.has_key?(content, key),
      do: Map.put(state, key, to_string(content[key])),
      else: state
  end

  # ── Merkle entries root (RFC 9162-style; mirrors merkle.ts) ─────────────────

  @doc """
  Merkle audit-log root (hex) over the ordered entries.
  Leaf hash: SHA-256(0x00 || canonicalJSON(entry)); node hash:
  SHA-256(0x01 || left || right); empty root: 64 hex zeros.
  """
  def entries_root(entries) do
    entries
    |> order_entries()
    |> Enum.map(fn e -> :crypto.hash(:sha256, <<0>> <> Security.canonical_json(e)) end)
    |> root_of()
    |> Base.encode16(case: :lower)
  end

  defp root_of([]), do: <<0::size(256)>>
  defp root_of([leaf]), do: leaf

  defp root_of(leaves) do
    n = length(leaves)
    mid = Bitwise.bsl(1, bit_width(n - 1) - 1)
    {left, right} = Enum.split(leaves, mid)
    :crypto.hash(:sha256, <<1>> <> root_of(left) <> root_of(right))
  end

  # Number of bits needed to represent n (n >= 1).
  defp bit_width(n), do: bit_width(n, 0)
  defp bit_width(0, acc), do: acc
  defp bit_width(n, acc), do: bit_width(Bitwise.bsr(n, 1), acc + 1)

  # ── base64url helpers ───────────────────────────────────────────────────────

  defp b64u(bytes), do: Base.url_encode64(bytes, padding: false)
  defp unb64u(s), do: Base.url_decode64!(s, padding: false)
end

File: python/ltx/src/interplanet_ltx/_registers.py
"""
interplanet_ltx._registers — Question and Action registers
Story 69.4 — LTX-SPECIFICATION.md §9/§10, LTX-SECURITY.md §9.5/§9.6.
Mirrors typescript/ltx/src/registers.ts.

Registers are deterministic reductions over the signed, append-only audit
log: entries ordered by (timestamp, nodeId, seq); object-level conflicts
resolve by highest object version, then lexicographically lowest editor
nodeId (§8.2). Entries are plain dicts using the same camelCase field names
as the wire format.
"""

import base64
from typing import Any, Dict, List, Optional

from ._security import canonical_json

ENTRY_PREFIX = {
    'question': 'QST', 'question_response': 'QST',
    'action': 'ACT', 'action_update': 'ACT',
    'amendment': 'AMD', 'state_transition': 'STA',
    'merge_snapshot': 'MRG', 'decision': 'DEC',
}

_ACTION_STATUSES = ('PROPOSED', 'ACCEPTED', 'REJECTED', 'DONE')

# Ed25519 DER wrapping constants (same idiom as _merkle.py / _security.py)
_PKCS8_HEADER = bytes.fromhex('302e020100300506032b657004220420')
_SPKI_HEADER = bytes.fromhex('302a300506032b6570032100')


def _b64u_decode(s: str) -> bytes:
    pad = 4 - len(s) % 4
    if pad != 4:
        s = s + '=' * pad
    return base64.urlsafe_b64decode(s)


def _sign(data: bytes, private_key_b64: str) -> str:
    from cryptography.hazmat.primitives.serialization import load_der_private_key
    raw_seed = _b64u_decode(private_key_b64)
    priv = load_der_private_key(_PKCS8_HEADER + raw_seed, password=None)
    sig = priv.sign(data)
    return base64.urlsafe_b64encode(sig).rstrip(b'=').decode()


def _verify(data: bytes, sig_b64: str, nik: Dict[str, Any]) -> bool:
    from cryptography.hazmat.primitives.serialization import load_der_public_key
    try:
        raw_pub = _b64u_decode(nik['publicKey'])
        pub = load_der_public_key(_SPKI_HEADER + raw_pub)
        pub.verify(_b64u_decode(sig_b64), data)
        return True
    except Exception:
        return False


def create_register_entry(entry_type: str, content: Dict[str, Any],
                          session_id: str, node_id: str, seq: int,
                          timestamp: str, private_key_b64: str,
                          entry_id: Optional[str] = None) -> Dict[str, Any]:
    """Create a signed register entry (LTX-SECURITY.md §9.5)."""
    eid = entry_id or f"{ENTRY_PREFIX[entry_type]}-{node_id}-{seq}"
    unsigned = {
        'entryId': eid,
        'sessionId': session_id,
        'nodeId': node_id,
        'seq': seq,
        'type': entry_type,
        'content': content,
        'timestamp': timestamp,
    }
    sig = _sign(canonical_json(unsigned).encode('utf-8'), private_key_b64)
    entry = dict(unsigned)
    entry['sig'] = sig
    return entry


def verify_register_entry(entry: Dict[str, Any],
                          key_cache: Dict[str, Any]) -> Dict[str, Any]:
    """Verify entry signature against key_cache[entry['nodeId']] -> NIK."""
    if not entry or not entry.get('sig'):
        return {'valid': False, 'reason': 'missing_sig'}
    nik = key_cache.get(entry.get('nodeId'))
    if not nik:
        return {'valid': False, 'reason': 'key_not_in_cache'}
    unsigned = {k: v for k, v in entry.items() if k != 'sig'}
    ok = _verify(canonical_json(unsigned).encode('utf-8'), entry['sig'], nik)
    return {'valid': True} if ok else {'valid': False, 'reason': 'signature_invalid'}


def order_entries(entries: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """De-duplicate by (nodeId, seq); sort by (timestamp, nodeId, seq) (§8.2)."""
    seen: Dict[str, Dict[str, Any]] = {}
    for e in entries:
        key = f"{e.get('nodeId')} {e.get('seq')}"
        if key not in seen:
            seen[key] = e
    return sorted(seen.values(),
                  key=lambda e: (e.get('timestamp', ''), e.get('nodeId', ''), e.get('seq', 0)))


def _wins(inc_version: int, inc_editor: str, cur_version: int, cur_editor: str) -> bool:
    if inc_version != cur_version:
        return inc_version > cur_version
    return inc_editor < cur_editor


def reduce_questions(entries: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Reduce question register state (LTX-SPECIFICATION.md §9.4). Pure."""
    by_id: Dict[str, Dict[str, Any]] = {}
    winners: Dict[str, Dict[str, Any]] = {}
    superseded: List[str] = []

    for e in order_entries(entries):
        c = e.get('content', {})
        if e.get('type') == 'question':
            qid = e['entryId']
            if qid in by_id:
                superseded.append(e['entryId'])
                continue
            winners[qid] = {'version': 1, 'editor': e['nodeId'], 'entryId': e['entryId']}
            q: Dict[str, Any] = {'qid': qid, 'text': str(c.get('text', '')),
                                 'submitter': e['nodeId']}
            if 'urgency' in c:
                q['urgency'] = str(c['urgency'])
            if 'intendedWindow' in c:
                q['intendedWindow'] = str(c['intendedWindow'])
            q['status'] = 'OPEN'
            q['version'] = 1
            by_id[qid] = q
        elif e.get('type') == 'question_response':
            qid = str(c.get('qid', ''))
            q = by_id.get(qid)
            if q is None:
                superseded.append(e['entryId'])
                continue
            version = int(c.get('version', q['version'] + 1))
            current = winners.get(qid)
            if current and not _wins(version, e['nodeId'], current['version'], current['editor']):
                superseded.append(e['entryId'])
                continue
            if current and current['entryId'] != q['qid']:
                superseded.append(current['entryId'])
            winners[qid] = {'version': version, 'editor': e['nodeId'], 'entryId': e['entryId']}
            q = dict(q)
            q['status'] = 'WITHDRAWN' if c.get('status') == 'WITHDRAWN' else 'ANSWERED'
            if 'response' in c:
                q['response'] = str(c['response'])
            q['responder'] = e['nodeId']
            q['version'] = version
            by_id[qid] = q
    return {'byId': by_id, 'superseded': superseded}


def reduce_actions(entries: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Reduce action register state (LTX-SPECIFICATION.md §10.2). Pure."""
    by_id: Dict[str, Dict[str, Any]] = {}
    winners: Dict[str, Dict[str, Any]] = {}
    superseded: List[str] = []

    for e in order_entries(entries):
        c = e.get('content', {})
        if e.get('type') == 'action':
            aid = e['entryId']
            if aid in by_id:
                superseded.append(e['entryId'])
                continue
            winners[aid] = {'version': 1, 'editor': e['nodeId'], 'entryId': e['entryId']}
            a: Dict[str, Any] = {'aid': aid, 'description': str(c.get('description', ''))}
            if 'owner' in c:
                a['owner'] = str(c['owner'])
            if 'dueTimeUTC' in c:
                a['dueTimeUTC'] = str(c['dueTimeUTC'])
            if 'originWindow' in c:
                a['originWindow'] = str(c['originWindow'])
            a['status'] = 'PROPOSED'
            a['version'] = 1
            by_id[aid] = a
        elif e.get('type') == 'action_update':
            aid = str(c.get('aid', ''))
            a = by_id.get(aid)
            if a is None:
                superseded.append(e['entryId'])
                continue
            version = int(c.get('version', a['version'] + 1))
            current = winners.get(aid)
            if current and not _wins(version, e['nodeId'], current['version'], current['editor']):
                superseded.append(e['entryId'])
                continue
            if current and current['entryId'] != a['aid']:
                superseded.append(current['entryId'])
            winners[aid] = {'version': version, 'editor': e['nodeId'], 'entryId': e['entryId']}
            a = dict(a)
            a['status'] = c['status'] if c.get('status') in _ACTION_STATUSES else a['status']
            if 'description' in c:
                a['description'] = str(c['description'])
            if 'owner' in c:
                a['owner'] = str(c['owner'])
            if 'dueTimeUTC' in c:
                a['dueTimeUTC'] = str(c['dueTimeUTC'])
            a['version'] = version
            by_id[aid] = a
    return {'byId': by_id, 'superseded': superseded}


def emit_question_seeds(seeds: List[Dict[str, Any]], session_id: str,
                        node_id: str, start_seq: int, timestamp: str,
                        private_key_b64: str) -> List[Dict[str, Any]]:
    """Re-emit v3 plan question seeds as signed log entries at lock (§9.2)."""
    out = []
    seq = start_seq
    for seed in seeds:
        out.append(create_register_entry('question', seed, session_id,
                                         node_id, seq, timestamp, private_key_b64))
        seq += 1
    return out
